presto-main/src/main/java/com/facebook/presto/execution/DataDefinitionExecution.java

```diff
@@ -65,6 +68,21 @@ private DataDefinitionExecution(
         this.accessControl = requireNonNull(accessControl, "accessControl is null");
         this.stateMachine = requireNonNull(stateMachine, "stateMachine is null");
         this.parameters = parameters;
+
+        stateMachine.addStateChangeListener(state -> {
```

Comment:
I believe `beginCreateTable`/`beginInsert` etc. has to be rewritten once this is merged since the contract for `beginQuery` states it's triggered once execution of query starts.
`beginInsert` is implemented in hacky way. It's triggered in the end of optimization phase but it's still in optimizer...